src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/ScanDecoder.cs

```diff
@@ -0,0 +1,887 @@
+﻿// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using SixLabors.ImageSharp.Formats.Jpeg.Components;
+using SixLabors.Memory;
+
+namespace SixLabors.ImageSharp.Formats.Jpeg.PdfJsPort.Components
+{
+    internal class ScanDecoder
+    {
+        public const int FastBits = 9;
+
+        // bmask[n] = (1 << n) - 1
+        private static readonly uint[] Bmask = { 0, 1, 3, 7, 15, 31, 63, 127, 255, 511, 1023, 2047, 4095, 8191, 16383, 32767, 65535 };
+
+        // bias[n] = (-1 << n) + 1
+        private static readonly int[] Bias = { 0, -1, -3, -7, -15, -31, -63, -127, -255, -511, -1023, -2047, -4095, -8191, -16383, -32767 };
+
+        private readonly DoubleBufferedStreamReader stream;
+        private readonly PdfJsFrameComponent[] components;
+        private readonly ZigZag dctZigZag;
+        private readonly int restartInterval;
+        private readonly int componentIndex;
+        private readonly int componentsLength;
+        private readonly int spectralStart;
+        private readonly int spectralEnd;
+        private readonly int successiveHigh;
+        private readonly int successiveLow;
+
+        private int codeBits;
+        private uint codeBuffer;
+        private bool nomore;
+        private bool eof;
+        private byte marker;
+        private bool badMarker;
+        private long markerPosition;
+        private int todo;
```

Comment:
Would be nice to have better names *or* some comments about the semantics of less trivial variables.

Reply:
Will do!